### PR TITLE
fix for #17 - deleting values from an iter_dup db fails

### PR DESCRIPTION
**Pull request by rrichardson**
A test for the bug defined in : https://github.com/mozilla/lmdb-rs/issues/17

### CODE REVIEW

**Comment by rrichardson** [comment]
OK.. I fixed it, but I don't totally understand it.  Apparently 
`&mut data_val as *mut _`  
is not the same as 
`&mut data_val`
even though the parameter to which they are passed coerces &mut to *mut... 
¯\_(ツ)_/¯

**Comment by mykmelez** [comment]
> OK.. I fixed it, but I don't totally understand it. Apparently
> `&mut data_val as *mut _`
> is not the same as
> `&mut data_val`
> even though the parameter to which they are passed coerces &mut to *mut...

I suspect it's because the `&mut data_val as *mut _` expression returns a raw pointer to the `mut data_val` parameter in the closure—which is not the same as the outer `data_val: Option<ffi::MDB_val>` variable—and once the closure returns, its `data_val` parameter gets dropped, dangling the pointer.

It looks like the explicit coercion from `&mut data_val` to `*mut _` was intended to ensure that the return value of the closure has the same type as the default value of the `Option.unwrap_or()` call.

Unfortunately, it had the side-effect of masking this bug, which becomes visible if you modify the original code to merely call `unwrap()` and then remove the explicit coercion, at which point you get an error like this one:

```bash
error[E0597]: `data_val` does not live long enough
   --> src/transaction.rs:339:71
    |
339 |                                      data_val.map(|mut data_val| &mut data_val)
    |                                                                       ^^^^^^^-
    |                                                                       |      |
    |                                                                       |      `data_val` dropped here while still borrowed
    |                                                                       borrowed value does not live long enough
340 |                                              .unwrap()))
    |                                                      - borrowed value needs to live until here
```

File: src/transaction.rs [hunk 1]
```diff
@@ -414,6 +425,49 @@ mod test {
...
+            let mut cur = txn.open_ro_cursor(db).unwrap();
+            let iter = cur.iter_dup_of(b"key1");
+            let vals = iter.map(|(_,x)| x).collect::<Vec<_>>();
+            assert_eq!(vals, vec![b"val1", b"val2", b"val3"]);
+
+        }
+        txn.commit().unwrap();
+
+        let mut txn = env.begin_rw_txn().unwrap();
+        txn.del(db, b"key1", Some(b"val2")).unwrap();
```
Comment (mykmelez) [inline src/transaction.rs]: Could you also add a test that confirms that `txn.del(db, b"key1", None)` deletes the other two values?

**Comment by mykmelez** [comment]
(Or, more accurately: to confirm that the other branch of the conditional behaves as expected when the database is configured as _DUP_SORT_. There are other tests that call `RwTransaction.del()` with a `None` value for the `data` parameter, but not for a _DUP_SORT_ database.)

File: src/transaction.rs (hunk 1 above)
Comment (rrichardson) [reply src/transaction.rs]: Done.

**Pull request merged**